File: 9.clj
#!/usr/bin/env bb

(defn pairs [l]
  (map vector l (rest l)))

(def data
  (-> (slurp "in/9.txt")
    (str/split #"\n")
    (->> (map #(->>
                 (str/split % #"\s")
                 (map parse-long))))))

;; 1

(defn first-solution [ss]
  (if (apply = ss)
    (last ss)
    (let [xx
          (+ (last ss)
            (first-solution (map (fn [[f s]] (- s f)) (pairs ss))))]
      xx)))

(apply + (for [l data] (first-solution l)))

;; 2

(defn second-solution [ss]
  (println ss)
  (if (apply = ss)
    (first ss)
    (let [xx
          (- (first ss)
            (second-solution (map (fn [[f s]] (- s f)) (pairs ss))))]
      (println xx)
      xx)))

(apply + (for [l data] (second-solution l)))

File: 2025/9.py
#!/usr/bin/env python3
from itertools import islice
from heapq import heappush_max as push, heappop_max as pop
from collections import defaultdict

f = "in/9.txt"
f = "in/9.test.txt"

data = [[int(x) for x in line.split(",")] for line in open(f).readlines()]

areas = []
for i, [x, y] in enumerate(data):
    for j, [dx, dy] in enumerate(islice(data, i + 1), i + 1):
        h = abs(x - dx) + 1
        v = abs(y - dy) + 1
        push(areas, (h * v, [x, y], [dx, dy]))

print("first", pop(areas))

data.sort(key=lambda x: x[1])
filled = set()

by_x = defaultdict(list)
by_y = defaultdict(list)

for x, y in data:
    by_x[x].append((x, y))
    by_y[y].append((x, y))

# for y in range(max_y):
#     l = ["." for _ in range(max_x + 1)]
#     for x, _ in by_y[y]:
#         l[x] = "#"
#     print("".join(l))
